src/lib.rs

```diff
@@ -53,189 +34,67 @@ extern crate rayon;
-    ///
-    /// See https://aws.amazon.com/ec2/instance-types/ for details.
-    pub instance_type: String,
-
-    /// The private IP address of this host on its designated VPC.
-    pub private_ip: String,
+pub mod providers;
+use providers::{Launcher, MachineSetup};
 
-    /// The publicly accessible hostname of this host.
+/// A handle to an instance currently running as part of a tsunami.
+/// Run commands on the machine via [ssh](ssh::Session)
+#[derive(Default)]
+pub struct Machine<'tsunami> {
+    pub nickname: String,
     pub public_dns: String,
-
-    /// The publicly accessible IP address of this host.
     pub public_ip: String,
-}
-
-/// A template for a particular machine setup in a tsunami.
-pub struct MachineSetup {
-    instance_type: String,
-    ami: String,
-    username: String,
-    setup: Box<Fn(&mut ssh::Session) -> Result<(), Error> + Sync>,
-}
 
-impl MachineSetup {
-    /// Define a new template for a tsunami machine setup.
-    ///
-    /// The given AWS EC2 instance type will be used. Note that only [EC2 Defined Duration Spot
-    /// Instance types](https://aws.amazon.com/ec2/spot/pricing/) are allowed.
-    ///
-    /// The `setup` argument is called once for every spawned instances of this type with a handle
-    /// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
-    /// commands on the host in question.
-    ///
-    /// The new instance will start out in the state dictated by the Amazon Machine Image specified
-    /// in `ami`.
-    ///
-    /// ```rust
-    /// # use tsunami::MachineSetup;
-    /// MachineSetup::new("m5.large", "ami-e18aa89b", |ssh| {
-    ///     ssh.cmd("cat /etc/hostname").map(|out| {
-    ///         println!("{}", out);
-    ///     })
-    /// });
-    /// ```
-    pub fn new<F>(instance_type: &str, ami: &str, setup: F) -> Self
-    where
-        F: Fn(&mut ssh::Session) -> Result<(), Error> + 'static + Sync,
-    {
-        MachineSetup {
-            instance_type: instance_type.to_string(),
-            ami: ami.to_string(),
-            username: String::from("ec2-user"),
-            setup: Box::new(setup),
-        }
-    }
+    /// If `Some(_)`, an established SSH session to this host.
+    pub ssh: Option<ssh::Session>,
 
-    /// Set the username to SSH into this machine type as.
-    ///
-    /// Defaults to `ec2-user`.
-    pub fn as_user(mut self, username: &str) -> Self {
-        self.username = username.to_string();
-        self
-    }
+    // tie the lifetime of the machine to the Tsunami.
+    _tsunami: std::marker::PhantomData<&'tsunami ()>,
 }
 
 /// Use this to prepare and execute a new tsunami.
-///
-/// A tsunami consists of one or more [`MachineSetup`](struct.MachineSetup.html)s that will be
-/// spawned as EC2 spot instances. See
-/// [`TsunamiBuilder#add_set`](struct.TsunamiBuilder.html#method.add_set)) for how to construct a
-/// tsunami.
+/// Call [add](TsunamiBuilder::add) to add machines to the Tsunami, and
```

Comment:
I'd prefer if method calls and such here where surrounded by backticks to give code font.

Reply:
I *think* I got all of them?